lib/Dialect/ModArith/Conversions/ModArithToArith/ModArithToArith.cpp

```diff
@@ -14,23 +16,207 @@ namespace mod_arith {
+      : OpConversionPattern<AddOp>(context) {}
+
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      AddOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    ImplicitLocOpBuilder b(op.getLoc(), rewriter);
+
+    auto cmod = b.create<arith::ConstantOp>(modulusAttr(op));
+    auto add = b.create<arith::AddIOp>(adaptor.getLhs(), adaptor.getRhs());
+    auto remu = b.create<arith::RemUIOp>(add, cmod);
+
+    rewriter.replaceOp(op, remu);
+    return success();
+  }
+};
+
+struct ConvertSub : public OpConversionPattern<SubOp> {
+  ConvertSub(mlir::MLIRContext *context)
+      : OpConversionPattern<SubOp>(context) {}
+
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      SubOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    ImplicitLocOpBuilder b(op.getLoc(), rewriter);
+
+    auto cmod = b.create<arith::ConstantOp>(modulusAttr(op));
+    auto sub = b.create<arith::SubIOp>(adaptor.getLhs(), adaptor.getRhs());
+    auto add = b.create<arith::AddIOp>(sub, cmod);
+    auto remu = b.create<arith::RemUIOp>(add, cmod);
+
+    rewriter.replaceOp(op, remu);
+    return success();
+  }
+};
+
+struct ConvertMul : public OpConversionPattern<MulOp> {
+  ConvertMul(mlir::MLIRContext *context)
+      : OpConversionPattern<MulOp>(context) {}
+
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      MulOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    ImplicitLocOpBuilder b(op.getLoc(), rewriter);
+
+    auto cmod = b.create<arith::ConstantOp>(modulusAttr(op, true));
+    auto lhs =
+        b.create<arith::ExtUIOp>(modulusType(op, true), adaptor.getLhs());
+    auto rhs =
+        b.create<arith::ExtUIOp>(modulusType(op, true), adaptor.getRhs());
+    auto mul = b.create<arith::MulIOp>(lhs, rhs);
+    auto remu = b.create<arith::RemUIOp>(mul, cmod);
+    auto trunc = b.create<arith::TruncIOp>(modulusType(op), remu);
+
+    rewriter.replaceOp(op, trunc);
+    return success();
+  }
+};
+
+struct ConvertMac : public OpConversionPattern<MacOp> {
+  ConvertMac(mlir::MLIRContext *context)
+      : OpConversionPattern<MacOp>(context) {}
+
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      MacOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    ImplicitLocOpBuilder b(op.getLoc(), rewriter);
+
+    auto cmod = b.create<arith::ConstantOp>(modulusAttr(op, true));
+    auto x = b.create<arith::ExtUIOp>(modulusType(op, true),
+                                      adaptor.getOperands()[0]);
+    auto y = b.create<arith::ExtUIOp>(modulusType(op, true),
+                                      adaptor.getOperands()[1]);
+    auto acc = b.create<arith::ExtUIOp>(modulusType(op, true),
+                                        adaptor.getOperands()[2]);
```

Comment:
just for later - the OpAdaptor comes with named accessors for the operands (like adaptor.getLhs(), adaptor.getRhs(), adapator.getAcc())